src/xml2csv.py

```diff
@@ -1,24 +1,89 @@
-from bs4 import BeautifulSoup
 import csv
+import gc
+import json
+import os
 import sys
+from datetime import datetime
+from tempfile import TemporaryDirectory
+
+import numpy as np
+from bs4 import BeautifulSoup
 
 XML_FILE_PATH = sys.argv[1]
 CSV_FILE_PATH = sys.argv[2]
 
-xml_file = open(XML_FILE_PATH, 'rb').read()
-xml_soup = BeautifulSoup(xml_file, 'lxml-xml')
-
-variables_file = open('data/datasets_format.html', 'rb').read()
-variables_soup = BeautifulSoup(variables_file, 'lxml')
-variables = variables_soup.select('.tabela-2 td')
-fieldnames = ['ideDocumento'] + \
-    [var.text.strip() for index, var in enumerate(variables) if index % 3 == 0]
-
-with open(CSV_FILE_PATH, 'w') as csv_file:
-    writer = csv.DictWriter(csv_file, fieldnames=fieldnames)
-    writer.writeheader()
-    records = xml_soup.find_all('DESPESA')
-    for record in records:
-        record_attributes = \
-            dict([(attr.name, attr.string) for attr in record.contents])
-        writer.writerow(record_attributes)
+field_names = ['ideDocumento']
+tmp_files = []
+
+
+def output(msg, **kwargs):
+    now = datetime.strftime(datetime.now(), '%Y-%m-%d %H:%M:%S')
+    return print(now, msg, **kwargs)
+
+# get extra field names
+output('Fetching CSV headers')
+with open('data/datasets_format.html', 'rb') as file_handler:
+    parsed = BeautifulSoup(file_handler.read(), 'lxml')
+    for row in parsed.select('.tabela-2 tr'):
+        try:
+            field_names.append(row.select('td')[0].text.strip())
+        except IndexError:
+            pass
+    del(parsed)
+    gc.collect()
```

Comment:
I've read about it but I'm not sure about how effective this actually is. 

Even after running `gc.collect()` my memory monitor still highlights a high memory usage. In spite of that I'm still able to allocate memory for other stuff.

My guess is that it might let the OS know that some blocks could be freed if needed and the OS handles it — but it's just a guess (and I haven't read the full documentation on the `gc` module either… so shame on me).